Group access to diagrams and folders

TanVD opened the pull request:
Now we support only one owner for each item. Also at some items owner even not stated. We want many to many relationship between folders and diagram and users. Also we need simple instrument in client to add someone to access group for selected item (diagram or folder).

TanVD committed:
Add set of owners in folder and updateFolder().

Replace string owner in Folder with set of owners. Need it for collaboration
and security purposes. Add updateFolder() operation in db-diagram-service.
Also need it for collaboration purposes.

TanVD committed:
Add sharing of folders.

Add many to many relation between folders. Also change thrift interfaces to support
changes. Also includes a lot of fixes and bugs tests, which were created while working
on this commit.

TanVD committed:
Sharing of folders.

Add One-To-Many to Many-To-Many relation mapping in db-diagram-service. Also
implements share other, more appropriate, way

TanVD committed:
Add README.md (russian).

Add README.md for db-diagram-service with full description of service and
of realization.

jzuken commented on db-services/db-diagram-service/README.md:
```diff
@@ -0,0 +1,87 @@
+её как надстройку над одним из сервисов инфраструктуры или как новый инфраструктурный сервис.
+
+ER диаграмма схемы данных - есть в архитектурном проекте (db-services-implementation-scheme), ER диаграмма логической схемы данных, предоставляемой в рамках одной
+сессии, так же там представлена (db-services-user-session-scheme).
```
"так же" -> "также"

jzuken replied:
also we better explicitly reference the UML/Architecture-WMP.vpp file here

jzuken commented on db-services/db-diagram-service/README.md:
```diff
@@ -0,0 +1,87 @@
+# Общие положения о разделении сервисов первой линии и сервисов инфраструктуры.
```
there should be no punctuation mark at the end of the header

jzuken commented on db-services/db-diagram-service/README.md:
```diff
@@ -0,0 +1,87 @@
+простейший CRUD, а не перегружаться логикой.
+
+Отображением внешнего представления во внутреннее и обратно занимаются сразу несколько классов. Во-первых есть отдельный класс RelationMapping, в нём содержатся методы дозаполнения класса
+Folder (например, догрузука родителей). Выделены они отдельно в связи со своей сложностью и необходимостью использования дополнительной информации. Все остальные методы перевода не слишком 
```
"догрузука". nice.
![image](https://cloud.githubusercontent.com/assets/705678/20037021/759a7d1c-a429-11e6-8380-4904ff2b315d.png)

jzuken commented on db-services/db-diagram-service/src/main/java/com/qreal/wmp/db/diagram/mapping/RelationMapping.java:
```diff
@@ -0,0 +1,95 @@
+            folder = loadFolderParent(folder);
+        }
+
+        for (Object child : folder.getChildrenFolders().toArray()) {
```
this code looks messy with all these casts. may be something like
`folder.setChildrenFolders(folder.getChildrenFolders().stream().map(child -> loadParents(child)).collect(Collectors.toSet()));`
?

jzuken commented on db-services/db-diagram-service/src/main/java/com/qreal/wmp/db/diagram/mapping/RelationMapping.java:
```diff
@@ -0,0 +1,95 @@
+        try {
+            loadedFolder = diagramDao.getFolder(folder.getId());
+        } catch (NotFoundException e) {
+            //logger.error("Consistency violation", e);
```
just do nothing?

jzuken commented on db-services/db-diagram-service/src/main/java/com/qreal/wmp/db/diagram/model/Folder.java:
```diff
@@ -69,13 +86,19 @@ public Folder(TFolder tFolder) {
             childrenFolders = tFolder.getChildrenFolders().stream().map(Folder::new).collect(Collectors.toSet());
         }
 
+        for (Object child : childrenFolders.toArray()) {
```
this will definitely look better with streams and map

jzuken commented on db-services/db-diagram-service/src/main/java/com/qreal/wmp/db/diagram/server/DiagramDbServiceHandler.java:
```diff
@@ -98,6 +131,63 @@ public TFolder getFolderTree(String username) throws TNotFound {
+    }
+
+    @Override
+    public void shareFolderTo(String username, TFolder tFolderToShare) throws TException {
```
shouldn't we use some kind of IDs for addressing users instead of just user names? but I guess that's a task for future pull-requests

TanVD replied:
Well, we use logins. It is unique id for every user. Why we do need something else?

jzuken replied:
idk, I guess we better identify them by emails or OpenID like all other services do. but all that is not very relevant right now

jzuken commented on db-services/db-diagram-service/src/main/java/com/qreal/wmp/db/diagram/server/DiagramDbServiceHandler.java:
```diff
@@ -98,6 +131,63 @@ public TFolder getFolderTree(String username) throws TNotFound {
+        Set<String> owners = folder.getOwners();
+        owners.add(username);
+        folder.setOwners(owners);
+        return folder;
```
this code looks weird. this `folder` got here by reference and therefore is already changed outside this method, isn't it?

TanVD replied:
Yes, but it's bad practice to implicitly update variable passed by reference in a method. So i explicitly return it.

jzuken replied:
No, it's a good practice to mark arguments that are not supposed to be changed as `final`. If the argument is not final, it could be changed by design.

jzuken commented on db-services/db-diagram-service/src/test/java/com/qreal/wmp/db/diagram/dao/DaoFolderBugsTest.java:
```diff
@@ -0,0 +1,91 @@
+        Folder testFolder = new Folder("testFolder", "testUser");
+        Folder testFolder1 = new Folder("testFolder1", "testUser");
+        testFolder.getChildrenFolders().add(testFolder1);
+
```
more empty lines for the god of empty lines!

jzuken commented on db-services/db-diagram-service/src/test/java/com/qreal/wmp/db/diagram/dao/DaoFolderRelationMappingTest.java:
```diff
@@ -0,0 +1,171 @@
+        Folder gotFolder = mapper.convertTFolder(testTFolder1);
+
+        assertThat(gotFolder).isEqualTo(testFolder1);
+        //assertThat(gotFolder.getParentFolders()).isEqualTo(testFolder1.getParentFolders());
```
should it really be here?

jzuken commented on db-services/startDb.sh:
```diff
@@ -0,0 +1,6 @@
+cd db-robot-store-service
```
what do we need this for?

TanVD replied:
Fast way to start db-services. Only for comfort. I don't see simple way to do it with maven.

jzuken commented:
so basically one can share a folder only from open or save dialogs? that seems strange, may be move it into separate menu next to file? anyway we need a place to put all collaboration stuff that comes later

TanVD commented:
I think we need totally new file explorer. Totally new concept of it. Adding new menu buttons for every new action not sounds like a good idea.

TanVD committed:
Style fixes (procedure to functional style)

Fixing few style violations. Also use functional style where it is appropriate.

TanVD committed:
Remove README.

Remove README. Fix contraversial function.

TanVD committed:
Merge branch 'master' into diagram-groups

TanVD committed:
Merge branch 'master' into diagram-groups

TanVD committed:
Merge master fix.

Add compiler changes. Fix .gitignore.

TanVD committed:
Add README.md

Add README.md for db-services and db-diagram-service.

TanVD committed:
Merge branch 'master' into diagram-groups

TanVD committed:
Delete unnecessary test

Delete test that we do not need since we have GUI tests.

TanVD committed:
Delete unused file.

Delete unused thrift binary.
